File: src/main/java/editor/database/symbol/HalfColorSymbol.java
package editor.database.symbol;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import editor.database.attributes.ManaType;

/**
 * This class represents half of a colored mana symbol.
 *
 * @author Alec Roelke
 */
public class HalfColorSymbol extends ManaSymbol
{
    /**
     * Map of color onto its corresponding half-symbol.
     */
    public static final Map<ManaType, HalfColorSymbol> SYMBOLS = Collections.unmodifiableMap(
            Arrays.stream(ManaType.values()).collect(Collectors.toMap(Function.identity(), HalfColorSymbol::new)));

    /**
     * Get the HalfColorSymbol corresponding to the given String.
     *
     * @param col String to get the symbol for
     * @return the HalfColorSymbol corresponding to the given color string
     * @throws IllegalArgumentException if the string does not describe a half color symbol
     */
    public static HalfColorSymbol parseHalfColorSymbol(String col) throws IllegalArgumentException
    {
        return tryParseHalfColorSymbol(String.valueOf(col.charAt(1))).orElseThrow(() -> new IllegalArgumentException('"' + col + "\" is not a half color symbol"));
    }

    /**
     * Get the HalfColorSymbol corresponding to the given String.
     *
     * @param col String to get the symbol for
     * @return the HalfColorSymbol corresponding to the given color string, or null if no such
     * symbol exists
     */
    public static Optional<HalfColorSymbol> tryParseHalfColorSymbol(String col)
    {
        if (col.length() == 2 && Character.toUpperCase(col.charAt(0)) == 'H')
            return Optional.ofNullable(SYMBOLS.get(ManaType.tryParseManaType(col.charAt(1))));
        return Optional.empty();
    }

    /**
     * This HalfColorSymbol's color.
     */
    private final ManaType color;

    /**
     * Create a new HalfColorSymbol.
     *
     * @param color color of the new HalfColorSymbol
     */
    private HalfColorSymbol(ManaType color)
    {
        super("half_" + color.toString().toLowerCase() + "_mana.png", "H" + String.valueOf(color.shorthand()), 0.5);
        this.color = color;
    }

    /**
     * {@inheritDoc}
     * This HalfColorSymbol's weights are 0.5 for its color and 0 for all the others.
     */
    @Override
    public Map<ManaType, Double> colorIntensity()
    {
        return createIntensity(new ColorIntensity(color, 0.5));
    }

    @Override
    public int compareTo(ManaSymbol o)
    {
        if (o instanceof HalfColorSymbol)
            return color.colorOrder(((HalfColorSymbol)o).color);
        else
            return super.compareTo(o);
    }
}

File: src/main/scala/editor/database/symbol/HalfColorSymbol.scala
package editor.database.symbol

import editor.database.attributes.ManaType

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

/**
 * Companion to [[HalfColorSymbol]] containing all possible [[HalfColorSymbol]]s and methods for parsing them from strings.
 * @author Alec Roelke
 */
object HalfColorSymbol {
  @deprecated val SYMBOLS = HalfColorSymbolGenerator.values.asJava
  @deprecated def tryParseHalfColorSymbol(s: String) = HalfColorSymbolGenerator.parse(s).toJava
  @deprecated def parseHalfColorSymbol(s: String) = HalfColorSymbolGenerator.values(ManaType.parseManaType(s))
}

/**
 * Mana symbol representing half of a specific type of mana (not to be confused with the generic 1/2 mana symbol, which is a [[StaticSymbol]]). Its
 * color intensity is 0.5 for its type of mana, and 0 for all others, as it is paid for by only half of that type of mana, and its text
 * representation is an H followed by its [[ManaType]]'s shorthand.
 * 
 * @constructor create a new half-mana symbol
 * @param color type of mana the symbol represents
 * 
 * @author Alec Roelke
 */
class HalfColorSymbol private[symbol](private val color: ManaType) extends ManaSymbol(s"half_${color.toString.toLowerCase}_mana.png", s"H${color.shorthand}", 0.5) {
  override def colorIntensity = ManaSymbol.createIntensity(Map(color -> 0.5))

  override def compareTo(o: ManaSymbol) = o match {
    case h: HalfColorSymbol => color.compareTo(h.color)
    case _ => super.compareTo(o)
  }
}